compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt

```diff
@@ -0,0 +1,424 @@
+
+                if (!wasDragInside && newIsDragInside) {
+                    handler.onDragStart(
+                        calculateOffset(
+                            componentCoordinates,
+                            newWindowDragCoordinates
+                        )
+                    )
+                }
+
+                if (wasDragInside && !newIsDragInside) {
+                    handler.onDragCancel()
+                }
+
+                if (newIsDragInside) {
+                    handler.onDrag(componentCoordinates.windowToLocal(newWindowDragCoordinates))
+                }
+            }
+            windowDragCoordinates = newWindowDragCoordinates
+        },
+        // notify components on window border drag exited window
+        onDragExit = {
+            for ((_, handler) in handlers) {
+                val componentCoordinates = handler.componentCoordinates
+                val oldDragCoordinates = windowDragCoordinates
+                val wasDragInside =
+                    isExternalDragInsideComponent(componentCoordinates, oldDragCoordinates)
+                if (wasDragInside) {
+                    handler.onDragCancel()
+                }
+            }
+            windowDragCoordinates = null
+        },
+        // notify all components under the pointer that drop happened
+        onDrop = {
+            var anyDrops = false
+            for ((_, handler) in handlers) {
+                if (isExternalDragInsideComponent(
+                        handler.componentCoordinates,
+                        windowDragCoordinates
+                    )
+                ) {
+                    handler.onDrop(it)
+                    anyDrops = true
+                }
+            }
+            windowDragCoordinates = null
+            // tell swing whether some components accepted the drop
+            return@AwtWindowDragTargetListener anyDrops
+        }
+    )
+
+    init {
+        addDropTargetListener(dragTargetListener)
+    }
+
+    override fun setActive(isActive: Boolean) {
+        super.setActive(isActive)
+        if (!isActive) {
+            windowDragCoordinates = null
+        }
+    }
+
+    /**
+     * Subscribes on drag events for [window].
+     * If drag is going and component is under pointer [onDragStart] can be called synchronously.
+     *
+     * @param componentCoordinates coordinates of the component used to properly detect when drag entered/exited component
+     * @return handler id that can be used later to remove subscription using [stopDragHandling]
+     */
+    fun installComponentDragHandler(
+        componentCoordinates: LayoutCoordinates,
+        onDragStart: (Offset) -> Unit,
+        onDrag: (Offset) -> Unit,
+        onDragCancel: () -> Unit,
+        onDrop: (DropData) -> Unit
+    ): Int {
+        isActive = true
+
+        handlers[idsCounter] = ComponentDragHandler(componentCoordinates, onDragStart, onDrag, onDragCancel, onDrop)
```

Comment:
Everything here goes in one thread (AWT), because of that idsCounter is not atomic

Reply:
Yes, `idsCounter` itself doesn't need to be atomic, but it's safer if the update is atomic in the thread, like 
```
val nextId = idsCounter++
handlers[nextId] = ...

return idsCounter
```
because who knows what onDragStart will do. It could for whatever reason call `installComponentDragHandler` itself.

It's really unlikely, so just to be on the really safe side.